crates/goose-cli/src/commands/term.rs

```diff
@@ -0,0 +1,243 @@
+function gt; {goose_bin} term run $argv; end
+
+# Log commands to goose
+function goose_preexec --on-event fish_preexec
+    string match -q -r '^goose term' -- $argv[1]; and return
+    string match -q -r '^gt($|\s)' -- $argv[1]; and return
+    {goose_bin} term log "$argv[1]" 2>/dev/null &
+end"#
+            )
+        }
+        "powershell" | "pwsh" => {
+            format!(
+                r#"$env:GOOSE_TERMINAL_ID = "{terminal_id}"
+function gt {{ & '{goose_bin}' term run @args }}
+
+# Log commands to goose
+Set-PSReadLineKeyHandler -Chord Enter -ScriptBlock {{
+    $line = $null
+    [Microsoft.PowerShell.PSConsoleReadLine]::GetBufferState([ref]$line, [ref]$null)
+    if ($line -notmatch '^goose term' -and $line -notmatch '^gt($|\s)') {{
+        Start-Job -ScriptBlock {{ & '{goose_bin}' term log $using:line }} | Out-Null
+    }}
+    [Microsoft.PowerShell.PSConsoleReadLine]::AcceptLine()
+}}"#
+            )
+        }
+        _ => {
+            return Err(anyhow!(
+                "Unsupported shell: {}. Supported shells: bash, zsh, fish, powershell",
+                shell
+            ));
+        }
+    };
+
+    println!("{}", script);
+    Ok(())
+}
+
+/// Handle `goose term log <command>` - log a shell command to the database
+pub async fn handle_term_log(command: String) -> Result<()> {
+    let terminal_id = std::env::var("GOOSE_TERMINAL_ID")
+        .map_err(|_| anyhow!("GOOSE_TERMINAL_ID not set. Run 'goose term init <shell>' first."))?;
+
+    let session_name = format!("{}{}", TERMINAL_SESSION_PREFIX, terminal_id);
+    let working_dir = std::env::current_dir()?;
+
+    ensure_terminal_session(session_name.clone(), working_dir.clone()).await?;
+    SessionManager::add_shell_command(&session_name, &command, &working_dir).await?;
+
+    Ok(())
+}
+
+/// Handle `goose term run <prompt>` - run a prompt in the terminal session
+pub async fn handle_term_run(prompt: String) -> Result<()> {
+    let terminal_id = std::env::var("GOOSE_TERMINAL_ID").map_err(|_| {
+        anyhow!(
+            "GOOSE_TERMINAL_ID not set.\n\n\
+             Add to your shell config (~/.zshrc or ~/.bashrc):\n    \
+             eval \"$(goose term init zsh)\"\n\n\
+             Then restart your terminal or run: source ~/.zshrc"
+        )
+    })?;
+
+    let session_name = format!("{}{}", TERMINAL_SESSION_PREFIX, terminal_id);
+    let working_dir = std::env::current_dir()?;
+
+    let session_id = match SessionManager::get_session(&session_name, false).await {
+        Ok(_) => {
+            SessionManager::update_session(&session_name)
+                .working_dir(working_dir)
+                .apply()
+                .await?;
+            session_name.clone()
+        }
+        Err(_) => {
+            ensure_terminal_session(session_name.clone(), working_dir).await?;
+            session_name.clone()
+        }
+    };
+
+    let commands = SessionManager::get_shell_commands_since_last_message(&session_id).await?;
```

Comment:
I'm not entirely sure I follow what we are fetching here. doesn't each session add at least one message to the conversation so isn't this then automatically empty?

in general though, adding shell commands to the session manager seems rather specific for this application. couldn't we just list the messages from session and exclude the <shell_history></shell_history> from that to get the same result? or if that doesn't work, maybe add a message content type?